lang/rs/bindings/cners/src/config.rs

```diff
@@ -0,0 +1,518 @@
+    pkt_api: Option<String>,
+    #[serde(default, rename = "no-metrics")]
+    no_metrics: bool,
+    #[serde(default, rename = "no-restapi")]
+    no_restapi: bool,
+    #[serde(default)]
+    cli: bool,
+    mode: Option<String>,
+    uds_path: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Thread {
+    group: Option<String>,
+    lport: Option<Vec<String>>,
+    description: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub(crate) struct Config<'a> {
+    application: Application,
+    defaults: Option<Defaults>,
+    umems: HashMap<String, UmemData<'a>>,
+    lports: HashMap<String, Lport<'a>>,
+    #[serde(rename = "lcore-groups")]
+    lcore_groups: Option<LcoreGroups>,
+    options: Option<Options>,
+    threads: Option<HashMap<String, Thread>>,
+}
+
+// Implement dummy Debug trait for xskdev_info_t.
+impl std::fmt::Debug for xskdev_info_t {
+    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Ok(())
+    }
+}
+
+impl<'a> Config<'a> {
+    pub fn load_config(jsonc_file: &str) -> Result<Config<'a>, CneError> {
+        let contents = fs::read_to_string(jsonc_file).map_err(|e| {
+            let msg = format!("Error reading file {} ", e);
+            CneError::ConfigError(msg)
+        })?;
+        let res = serde_json::from_str::<Config>(&contents).map_err(|e| {
+            let msg = format!("serde json error {} ", e);
+            CneError::ConfigError(msg)
+        });
+        res
+    }
+
+    #[cfg(test)]
+    pub fn get_config(cfg: &Config) -> Result<String, CneError> {
+        serde_json::to_string(cfg).map_err(|e| CneError::ConfigError(e.to_string()))
+    }
+
+    pub fn setup(&mut self) -> Result<(), CneError> {
+        self.setup_mem_pool()?;
+
+        self.setup_lports()?;
+
+        Ok(())
+    }
+
+    pub fn cleanup(&mut self) -> Result<(), CneError> {
+        self.cleanup_lports()?;
+
+        self.cleanup_mem_pool()?;
+
+        Ok(())
+    }
+
+    pub fn get_port_by_name(&self, port_name: &str) -> Result<Port, CneError> {
+        let lport = self.lports.get(port_name);
+
+        let ret = match lport {
+            Some(lport) => {
+                let ret = match lport.lport_id {
+                    Some(lport_id) => Ok(Port::new(lport_id)),
+                    None => {
+                        let msg = format!("Port name {} is not configured", port_name);
+                        Err(CneError::PortError(msg))
+                    }
+                };
+                ret
+            }
+            None => {
+                let msg = format!("Port name {} is not present in config", port_name);
+                Err(CneError::PortError(msg))
+            }
+        };
+
+        ret
+    }
+
+    fn setup_mem_pool(&mut self) -> Result<(), CneError> {
+        for (umem_name, umem) in &mut self.umems {
+            umem.name = umem_name.clone();
+            umem.bufcnt *= 1024;
+            umem.bufsz *= 1024;
+
+            if umem.rxdesc == 0 {
```

Comment:
Does rust require {} around single line if, while, for, ... statements?

Does Rust have a formatter? If the formatter add them it is OK.